### ISSUE TITLE
[css-nesting] Concern about combinatorial explosion

### ISSUE THREAD

**Issue by upsuper** [w3c/csswg-drafts#2881]
The most intuitive way to implement nesting rules is probably just expand them internally as their complete form, just like what preprocessors do nowadays.

But this has a problem that it means the number of selectors to match can grow exponentially as the nesting get deeper. This is less a concern when it was handled by preprocessors, because authors would see the exponential result from the generated file. If they see a generated CSS sized hundreds of megabytes or even gigabytes, they know there is a problem. But with nesting rules, they can hand to browsers a CSS file with just a handful of kilobytes which can be expanded to millions of rules, that would be problematic.

Consider something like
```css
.a1, .a2 {
&.b1, &.b2 {
&.c1, &.c2 {
&.d1, &.d2 {
&.e1, &.e2 {
&.f1, &.f2 {
&.g1, &.g2 {
&.h1, &.h2 {
&.i1, &.i2 {
&.j1, &.j2 {
&.k1, &.k2 {
&.l1, &.l2 {
&.m1, &.m2 {
&.n1, &.n2 {
&.o1, &.o2 {
&.p1, &.p2 {
&.q1, &.q2 {
&.r1, &.r2 {
&.s1, &.s2 {
&.t1, &.t2 {
&.u1, &.u2 {
&.v1, &.v2 {
&.w1, &.w2 {
&.x1, &.x2 {
&.y1, &.y2 {
&.z1, &.z2 {
```

It's only 300+ bytes, but can be expanded into 67M selectors. It's probably not good.

Potential solutions includes:
1. restrict the maximum nesting levels (e.g. at most 3 levels are allowed)
2. allow only a single complex selector, rather than a list in nested rules

**Comment by tabatkins** [comment]
Allowing selector lists is necessary; the combinatorial explosion is, in large part, the *point* of nesting.

I'm fine with allowing a max, tho 3 is a little low. Perhaps 5 or 6?

**Comment by upsuper** [comment]
I have a feeling that shortening selectors is also an important usefulness from nesting, so I don't completely buy that combinatorial explosion is *the* point, but I'm okay with that argument.

Do we have data on how many levels do people do nowadays with preprocessors? I assume two levels (one top level and one level nested) should cover majority of cases, and three levels should almost cover the rest. Do people use 5 or 6 levels a lot? With 6 level, you just need 10 selectors in each level to get 1M, which feels problematic still.

**Comment by jonjohnjohnson** [comment]
Even disallowing more than 6 levels won’t stop the possible explosion using large ‘:matches()’ lists at each of the possible 6 levels?

**Comment by upsuper** [comment]
I don't think anyone wants to expand `:matches()` at all, and not expanding it has a reasonable matching complexity as well, especially if we don't have to handle specificity. (The hard part of `:matches()` in implementation is invalidation, not matching.)

Nested rules are different. On the one hand, while `a1, a2 { & b1, & b2 { } }` is equivalent to `:matches(a1, a2) :matches(b1, b2)`, `a b { & c d { } }` means just `a b c d` and nothing else, unlike `:matches(a b) :matches(c d)`. On the other hand, for nested rules, the parent rule can appear in anywhere in child rule, for example you can say `a b { c d & { } }` which means `c d a b` (syntax is not going to be this way but that's the idea). This means rewriting nested rules into a single complex selector without repeating anything is basically impossible, so you would need to expand it anyway.

**Comment by upsuper** [comment]
(Specificity handling is also hard for `:matches()` in implementation I guess.)

**Comment by tabatkins** [comment]
> Nested rules are different.

No, they're not, you're just desugaring wrong. ^_^ When desugaring, you replace the `&` with a `:matches()` containing the parent selector

* `a b { & c d { } }` desugars to `:matches(a b) c d {}`. 
* `a1, a2 { & b1, & b2 { } }` desugars to `:matches(a1, a2) b1, :matches(a1, a2) b2 {}`.  (It so happens that this is equivalent to the wrong desugaring you gave.)
* `a b { c d & { } }` desugars to `c d :matches(a b)`.

> (Specificity handling is also hard for :matches() in implementation I guess.)

No, we resolved to use the simpler version of :matches() specificity - it's the specificity of the most specific argument, regardless of matching.

**Comment by upsuper** [comment]
> No, they're not, you're just desugaring wrong. ^_^ When desugaring, you replace the `&` with a `:matches()` containing the parent selector

So it's different from what preprocessors do currently? I don't think preprocessors are able to expand `c d :matches(a b)` to a selector pre-matches, are they?

Anyway, expanding them to `:matches` would only make them worse than `:matches`. Still, I don't believe you can expand any nested selector into a single complex selector. Consider for example, `a b { & c d { } e f & { } }`, what selector would you expand it to?

> No, we resolved to use the simpler version of :matches() specificity - it's the specificity of the most specific argument, regardless of matching.

Oh, okay, that's great. Then `:matches()` is probably not very hard to implement.

**Comment by tabatkins** [comment]
> I don't think preprocessors are able to expand c d :matches(a b) to a selector pre-matches, are they?

They certainly can, but it explodes combinatorially:

```
c d a b
c da b
c a d b
ca d b
a c d b
```

In cases like this (like Sass's @extend), they use some heuristics to avoid having to fully expand the possiblities, while making it highly likely that what they do expand to is useful, but that's not strictly necessary.

> Consider for example, a b { & c d { } e f & { } }, what selector would you expand it to?

Just run the desugaring twice, man:  `e f :matches(:matches(a b) c d)`.

**Comment by upsuper** [comment]
> Just run the desugaring twice, man: `e f :matches(:matches(a b) c d)`.

That's different, no? In your expanded form it requires all of the a, b, c, d, e, f to appear in the path somehow, while in the original form, only a and b must appear, while either c and d, or e and f need to also appear.

**Comment by tabatkins** [comment]
No, all of them definitely need to appear, in exactly the relationship I described. What makes you think you can skip any of c/d/e/f? Imagine what elements are matched at each level.

**Comment by Loirooriol** [comment]
From my understanding,

```css
a b {
  & c d { }
  e f & { }
}
```
should expand to
```css
:matches(a b) c d { }
e f :matches(a b) { }
```
Sure, the parent part is repeated but there is no combinatorial explosion.

The problem seems indeed when `&` is used multiple times in an inner selector:

```css
a1, a2 {
  & b1, & b2 {
    & c1, & c2 {}
  }
}
```

expands to

```css
:matches(:matches(a1, a2) b1, :matches(a1, a2) b2) c1,
:matches(:matches(a1, a2) b1, :matches(a1, a2) b2) c2 {}
```

but then I think implementations should just not do this, and reuse the parent part without actually duplicating it.

**Comment by upsuper** [comment]
> but then I think implementations should just not do this, and reuse the parent part without actually duplicating it.

I thought it is very tricky to implement nesting without actually expanding them. But after thinking about it further, I think there is a reasonable simple algorithm to implement this without expanding, so I'm going to close this issue.

I'm a bit concern about it's indication of promoting use of `:matches`, but that's probably a separate issue.

**Comment by upsuper** [comment]
After thinking a bit more, I think I was wrong in the previous comment, and I still cannot figure out a universal way to implement nesting rules efficiently without exponentially expanding the selectors.

If we can resolve to use one-element-a-time semantic in #2895, and also to require balanced nesting selector in #2896, and maybe disallowing `:not()` to contain `&`, there can be a decent approach to implement but I'm not completely sure.

**Comment by upsuper** [comment]
I guess anything in the latter half of the previous comment doesn't really matter. The only reasonable way to implement it seems to just be running matching exponentially... or alternatively, having a exponential memory usage and do stuff linearly.

**Comment by tabatkins** [comment]
Oh! I misread your rule, I didn't realize they were two rules nested in the single parent rule.

(Tangent: that's an invalid rule, btw. Can't have a rule like `e f & {}`, needs to be `@nest e f & {}`.)

So that expands to the two rules `:matches(a b) c d {} e f :matches(a b) {}`. Or if you wanted to use your comma-separated one from the last comment, `:matches(a b) c d, e f :matches(a b) {}`. 

What did you think was complicated about desugaring that?

**Comment by upsuper** [comment]
I don't think it's complicated to desugar that. The problem is the number of selectors after desugaring grow exponentially and I cannot figure out a reasonable algorithm for matching which neither time nor memory is exponential.

**Comment by tabatkins** [comment]
It doesn't grow at all if you use :matches() internally, per the desugaring.

**Comment by upsuper** [comment]
Could you desugar the following selectors with `:matches()`?
```css
a, a {
@nest & b, b & {
@nest & c, c & {
@nest & d, d & {
@nest & e, e & {
@nest & f, f & {
```

**Comment by tabatkins** [comment]
The innermost rule desugars down to:

`:matches(:matches(:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2))) d, d:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2)))) e, e :matches(:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2))) d, d:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2))))) f, f:matches(:matches(:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2))) d, d:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2)))) e, e :matches(:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2))) d, d:matches(:matches(:matches(a1, a2) b, b :matches(a1, a2)) c, c :matches(:matches(a1, a2) b, b :matches(a1, a2)))))`

The higher rules desugar to subsets of that.

**Comment by upsuper** [comment]
That's exponential to the length of the selectors written above, isn't it? That one has only 6 levels, so you can still list it. If I wrote 26 levels like [the previous comment](https://github.com/w3c/csswg-drafts/issues/2881#issue-338395788), maybe your desugared rules would get rejected by GitHub as a comment :D

**Comment by tabatkins** [comment]
Sure. Very deep nesting of selectors is uncommon in the first place, tho (based on preprocessor experience), and deep nesting of *selector lists* is even less common. When things are deeply nested, they commonly just do appending, like `.foo { & .bar { & .baz {...`, which is easy to handle.

That said, I'm still okay with a maximum nesting depth set to something reasonable, like 5 or 6.

**Comment by pygy** [comment]
Wouldn't a limit on the total length of the expanded selector list make more sense?

Not that nesting 6 levels deep is wise, but you a hostile style sheet could still do some damage by using a large lists of selectors at every level (oh, you've disabled JavaScript? How's *that*? [BAM, here comes a selector bomb that would have been unloaded synchronously by a JS script]).

**Comment by tabatkins** [comment]
The length of the selector list at a given level doesn't cause any explosion in the `:matches()`-based desugaring; the higher levels gets repeated more, but doubling the length of the top level just approximately doubles the length of the result. It's the nesting level that causes non-linear growth.

**Comment by pygy** [comment]
You still get a n^6 bang for your buck. That's a sweet payout.

Contrast that with an iframe bomb that has a linear server cost.

```
b, b b, b b b, b b b b, b b b b b, b b b b b b {
@nest & , b &, b & b, b b & b, b b b & b , b b b b & b, b b b b b & b, b b b b b b & {
@nest & , b &, b & b, b b & b, b b b & b , b b b b & b, b b b b b & b, b b b b b b & {
@nest & , b &, b & b, b b & b, b b b & b , b b b b & b, b b b b b & b, b b b b b b & {
@nest & , b &, b & b, b b & b, b b b & b , b b b b & b, b b b b b & b, b b b b b b & {
@nest & , b &, b & b, b b & b, b b b & b , b b b b & b, b b b b b & b, b b b b b b & {
}}}}}
```

Expands to a selector that is 1984790 characters long.

```JS
[...Array(5)].map(
  ()=>'& , b &, b & b, b b & b, b b b & b , b b b b & b, b b b b b & b, b b b b b b &'
).reduce(
  (r, v) => v.replace(/&/g, `is(${r})`),
  'b, b b, b b b, b b b b, b b b b b, b b b b b b'
).length // 1984790
```

Even if assuming that this example can be handled smartly by the engine because of the repetition of the `b`, you'd lose that if the attacker uses variation of the pattern with distinct selectors.

**Comment by argyleink** [comment]
I could see capping at 6 deep being reasonable. I agree it's uncommon to see nesting deeper than that. Does seem like a limit should be present though, just like with custom properties and the billion laughs attack it can produce without the limit.

**Comment by fantasai** [comment]
Agenda+ to discuss if we want to:

- specify a cap
- specify a minimum
- do nothing

**Comment by Loirooriol** [comment]
Following the same reasoning as in https://github.com/w3c/csswg-drafts/issues/3462, I'm fine with

 - Minimum in the spec
 - No explicit maximum in the spec
 - Implementation-defined maximum

**Comment by tabatkins** [comment]
Going ahead and tagging this for f2f discussion, as it seems good to hash this out.

**Comment by tabatkins** [comment]
Lol that is way higher than the minimum I'd add, so that's fine. (I was considering, like, 10 or something.)

**Comment by Loirooriol** [comment]
@sesse Correct me if I'm wrong, but my understanding of https://github.com/w3c/csswg-drafts/issues/8310#issuecomment-1383810771 is that Blink doesn't expand `&` so it's not affected by this problem. Couldn't other implementations use the same approach?

**Comment by mdubet** [comment]
WebKit expands `&` to `:is(parent_selector_list)`, which grows to something like `O(number_of_levels^average_number_of_nesting_selectors_at_each_level)` - which is not exponential on the number of levels I think?

Also, we indeed have a maximum rule nesting level, but not specifically for style rule nesting.

**Comment by Loirooriol** [comment]
This testcase has 25 levels:

```html
<!DOCTYPE html>
<style>
& e1a, & e1b {
  & e2a, & e2b {
    & e3a, & e3b {
      & e4a, & e4b {
        & e5a, & e5b {
          & e6a, & e6b {
            & e7a, & e7b {
              & e8a, & e8b {
                & e9a, & e9b {
                  & e10a, & e10b {
                    & e11a, & e11b {
                      & e12a, & e12b {
                        & e13a, & e13b {
                          & e14a, & e14b {
                            & e15a, & e15b {
                              & e16a, & e16b {
                                & e17a, & e17b {
                                  & e18a, & e18b {
                                    & e19a, & e19b {
                                      & e20a, & e20b {
                                        & e21a, & e21b {
                                          & e22a, & e22b {
                                            & e23a, & e23b {
                                              & e24a, & e24b {
                                                & e25a, & e25b {
                                                  background: lime;
                                                }
                                              }
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
</style>
<e1a><e2a><e3a><e4a><e5a><e6a><e7a><e8a><e9a><e10a><e11a><e12a><e13a><e14a><e15a><e16a><e17a><e18a><e19a><e20a><e21a><e22a><e23a><e24a><e25a>text
<script>
let t = performance.now();
document.body.offsetLeft;
document.body.append(performance.now() - t);
</script>
```

Blink takes ~10 seconds, WebKit crashes after a long freeze. So Blink is better but it still seems affected by this problem.

**Comment by emilio** [comment]
Live test-case: http://crisal.io/tmp/lots-of-nesting.html

Firefox Nightly takes ~4s on my machine fwiw :P

**Comment by Loirooriol** [comment]
I was wrongly getting few ms in Nightly, I have moved `let t = performance.now();` to the very top, then I get like 6s.

**Comment by tabatkins** [comment]
Note that Chrome has the same timing if each rule is a single selector wrapped in `:is()` (that is, `:is(& e1a, & e1b) {...}`, so the perf issues are introduced by something other than the branching itself. (This also means we can't base it on a complexity budget depending on the length of the selector lists.)

**Comment by tabatkins** [comment]
And Chrome has *different* timing if you collapse away the nesting. That is, `:is(a, b) { :is(c, d) {...}}` and `:is(a, b) :is(c, d) ... {...}` are at least a 10x difference in timing (with the 25 `:is()` functions, i got 0.8s vs 13s on my laptop). So it's still slow, but there's something else also contributing to the blowup.

**Comment by SebastianZ** [comment]
Just a tangential side note on this: When UAs hit a selector/nesting limit, their DevTools should give the authors a hint noting that.

Sebastian

**Comment by FremyCompany** [comment]
> And Chrome has _different_ timing if you collapse away the nesting. That is, `:is(a, b) { :is(c, d) {...}}` and `:is(a, b) :is(c, d) ... {...}` are at least a 10x difference in timing (with the 25 `:is()` functions, i got 0.8s vs 13s on my laptop). So it's still slow, but there's something else also contributing to the blowup.

That seems to indicate that there is a design choice in Chrome's implementation that is causing this to be an issue at levels lower than it could in theory. Maybe once that root cause is identified, it would be possible to agree on a very large limit and call it a day.

**Comment by mdubet** [comment]
We already have this kind of "not obvious for author" exponential expansion with `--var`. The css-variables spec warns about this, maybe the css-nesting spec could have a similar paragraph.

> To avoid this sort of attack, UAs must impose a UA-defined limit on the allowed length of the token stream that a [var()](https://drafts.csswg.org/css-variables-2/#funcdef-var) function expands into. If a var() would expand into a longer token stream than this limit, it instead makes the property it’s expanding into [invalid at computed-value time](https://drafts.csswg.org/css-variables-2/#invalid-at-computed-value-time). 

https://drafts.csswg.org/css-variables-2/#long-variables
